### PR TITLE
Feature request: Add a graphical display of logobject timelines #17

### LINKED ISSUE #17: Feature request: Add a graphical display of logobject timelines
Skjermbilde 2019-10-02 kl  11 14 49

This would make it easier to see how multiple logobjects aligns with each other.

### PR DESCRIPTION
**Pull request by robertbasti**
[//]: # (Request Template Witsml Explorer)

[//]: # (Thank you for contributing to WITSML Explorer! Before submitting this PR, please fill in the following template.)

## Fixes

This pull request fixes #17

## Description

 adds graphical view of logs

## Type of change

[//]: # (Mark any of the types of change that apply.)

* [ ] Bugfix
* [x] New feature (non-breaking change which adds functionality)
* [ ] Enhancement of existing functionality
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update

## Impacted Areas in Application

[//]: # (List general components of the application that this PR will affect)

* [x] Frontend
* [ ] API
* [ ] WITSML
* [ ] Other (please describe)

## Checklist:

[//]: # (Please tick all the boxes or remove the ones that aren't needed and explain why)

*Communication*
* [ ] I have made corresponding changes to the documentation
* [ ] PR affects application security

*Code quality*
* [x] I have self-reviewed my code
* [x] No new warnings are generated

*Test coverage*
* [ ] New code is covered by passing tests

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: Src/WitsmlExplorer.Api/Middleware/ErrorDetails.cs [hunk 1]
```diff
@@ -6,7 +6,6 @@ public class ErrorDetails
     {
         public int StatusCode { get; set; }
         public string Message { get; set; }
-
```
Comment (janburak) [inline Src/WitsmlExplorer.Api/Middleware/ErrorDetails.cs:9]: Unnecessary change

File: Src/WitsmlExplorer.Frontend/components/ContentView.tsx [hunk 2]
```diff
@@ -81,6 +83,8 @@ const ContentView = (): React.ReactElement => {
         setObjectView(true);
       } else if (currentSelected === selectedLogTypeGroup) {
         setView(<LogsListView />);
+      } else if (currentSelected === displayGraph) {
+        setView(<LogsGraph />);
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentView.tsx]: Is there a reason for using NavigationContext for switching to LogsGraph? Would it not be simpler to keep the `displayGraph` variable in LogsListView and do something like `return displayGraph ? <ContentTable/> : <LogsGraph logs={logs}/>;`

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+  const data: DataItem[] = [];
+  const categories: number[] = [];
+  const { navigationState } = useContext(NavigationContext);
+  const { selectedWellbore, selectedLogTypeGroup } = navigationState;
+
+  // dedicated colors for items containing specific names
+  const reservedColours: Map<string, string> = new Map([
+    ["#0000FF", "surface"],
+    ["#FF0000", "downhole"],
+    ["#FFFF00", "memory"]
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx:45]: The color for "memory" makes the white log name label unreadable.

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    if (numberOfCategories < 13) {
+      return numberOfCategories * 50;
+    }
+    return 600;
+  };
+
+  const randomColor = (name: string): string => {
+    let result = "";
+    reservedColours.forEach(function (value, key) {
+      if (name.includes(value)) {
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx]: Should be `name.toLowercase().includes(value)`

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+      }
+    });
+    // returns reserved colours
+    if (result !== "") {
+      return result;
+    }
+    return generateColor(name);
+  };
+
+  const generateColor = (name: string): string => {
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx]: The random colors make the graph look a bit chaotic. Should the same color be used for all "other" logs, or maybe the color palette should be narrowed down? Maybe discuss it with Asbjørn.

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+          x: [1, 2],
+          y: 0,
+          tooltip: [dimStart, dimEnd]
+        },
+        data: data
+      }
+    ]
+  };
+
+  useEffect(() => {
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx]: This useEffect and the one below seem to be unnecessary?

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+    });
+  };
+
+  const sortedLogs = getTableData().sort((a, b) => {
+    return a.start - b.start;
+  });
+
+  // calculates equal vertical distribution
+  // needs to be more tuned
+  const verticalAxisZoomMaxValue = () => {
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx:83]: Can possibly replace the following with, as an example, :
```
  const barHeight = 30;
  const spacing = 30;
  const verticalAxisZoomMaxValue = () => {
    return gridMaxHeight() / (categories.length * (barHeight + spacing)) * 100;
  };

  const gridMaxHeight = () => {
    const numberOfCategories = categories.length;
    if (numberOfCategories < 13) {
      return numberOfCategories * (barHeight + spacing);
    }
    return 600;
  };
```
It can be made more precise by including the spacing before the first item and after the last item.

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/ReactLogChart.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      // eslint-disable-next-line @typescript-eslint/no-unused-expressions
+      loading === true ? chart?.showLoading() : chart?.hideLoading();
+    }
+  }, [loading, theme]);
+
+  const { navigationState, dispatchNavigation } = useContext(NavigationContext);
+
+  const { selectedWell, selectedWellbore } = navigationState;
+
+  const onSelect = (log: LogObjectRow) => {
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/ReactLogChart.tsx]: Only `log.logObject`, so sending in a `LogObjectRow` is unnecessary.

File: Src/WitsmlExplorer.Frontend/components/ContentViews/table/Panel.tsx [hunk 9]
```diff
@@ -36,13 +39,14 @@ const Panel = (props: PanelProps) => {
     columns,
     expandableRows = false,
     downloadToCsvFileName = null,
-    stickyLeftColumns
+    stickyLeftColumns,
+    showGraph = false
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/table/Panel.tsx]: Given that `showGraph` is used by only one view, the related functionality should be removed from `Panel`, and the display graph button should be passed through `panelElements` instead.

File: Src/WitsmlExplorer.Frontend/components/Routing.tsx
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/Routing.tsx:1]: Changes to `Routing` are unnecessary.

File: Src/WitsmlExplorer.Frontend/components/ContentView.tsx (hunk 2 above)
Comment (janburak) [reply Src/WitsmlExplorer.Frontend/components/ContentView.tsx]: See CurveValuesView in https://github.com/equinor/witsml-explorer/pull/1995/files as an example.

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx (hunk 5 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx]: Discussion with Asbjørn is needed.

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx (hunk 3 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx:45]: Discussion with Asbjørn is needed.

File: Src/WitsmlExplorer.Frontend/components/ContentView.tsx (hunk 2 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentView.tsx:87]: I followed your instruction by creating a new navigation type NavigationType.SelectLogTypeGraph. For me it seems to more logical approach.

File: Src/WitsmlExplorer.Frontend/components/Routing.tsx
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/Routing.tsx:1]: fixed

File: Src/WitsmlExplorer.Frontend/components/ContentViews/table/Panel.tsx (hunk 9 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/table/Panel.tsx]: fixed

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/ReactLogChart.tsx (hunk 8 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/ReactLogChart.tsx]: fixed

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx (hunk 7 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx:83]: fixed

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx (hunk 6 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx]: fixed

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx (hunk 4 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx]: fixed

File: Src/WitsmlExplorer.Api/Middleware/ErrorDetails.cs (hunk 1 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Middleware/ErrorDetails.cs:9]: fixed

File: Src/WitsmlExplorer.Frontend/components/ContentView.tsx (hunk 2 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentView.tsx]: implemented, but needed to delete the new action and button

File: Src/WitsmlExplorer.Frontend/components/ContentViews/table/Panel.tsx (hunk 9 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/table/Panel.tsx]: later also removed that due to using switch on view instead of button

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+      }
+    });
+  }
+
+  const renderGanttItem: CustomSeriesRenderItem = (params: CustomSeriesRenderItemParams, api: CustomSeriesRenderItemAPI) => {
+    const itemIndex = api.value(dimItemIndex);
+    const start = api.coord([api.value(dimStart), itemIndex]);
+    const end = api.coord([api.value(dimEnd), itemIndex]);
+    let barLength = end[0] - start[0];
+    const barHeight = 30;
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx]: Duplicates `const barHeight` from line 82

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+    let x = start[0];
+    const y = end[1];
+    //assures minimum bar lenght to be visible in graph
+    const originalX = sortedLogs[itemIndex as number].start;
+    console.log(originalX);
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx]: Remove console.log

File: Src/WitsmlExplorer.Frontend/components/ContentView.tsx [hunk 12]
```diff
@@ -54,7 +55,8 @@ const objectViews: Partial<Record<ObjectType, ReactElement>> = {
 
 const ContentView = (): React.ReactElement => {
   const { navigationState } = useContext(NavigationContext);
-  const { selectedWell, selectedWellbore, selectedLogTypeGroup, selectedLogCurveInfo, selectedObjectGroup, selectedObject, selectedServer, currentSelected } = navigationState;
+  const { selectedWell, selectedWellbore, selectedLogTypeGroup, selectedLogCurveInfo, selectedObjectGroup, selectedObject, selectedServer, currentSelected, displayGraph } =
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentView.tsx]: `displayGraph` related code can be removed:
* lines 58, 86, 87 ContentView
* lines 21, 43 Routing
* lines 44, 73 navigationContext

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+        category: null
+      };
+    });
+  };
+
+  const sortedLogs = getTableData().sort((a, b) => {
+    return a.start - b.start;
+  });
+
+  // calculates equal vertical distribution
```
Comment (janburak) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx:80]: can probably remove the comment "// needs to be more tuned" now?

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx (hunk 10 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx]: fixed

File: Src/WitsmlExplorer.Frontend/components/ContentView.tsx (hunk 12 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentView.tsx]: fixed

File: Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx (hunk 13 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/Charts/LogsGraph.tsx:80]: removed

**Review by janburak — APPROVED** [review]
- [X] Read issue and PR
- [X] Pulled branch and manually tested
- [X] Verify that PR resolves issue
- [X] Reviewed the code
